webhook/resources/backingimage/validator.go

```diff
@@ -0,0 +1,114 @@
+	}
+
+	if len(backingImage.Spec.Checksum) != 0 {
+		if !util.ValidateChecksumSHA512(backingImage.Spec.Checksum) {
+			return werror.NewInvalidError(fmt.Sprintf("invalid checksum %v", backingImage.Spec.Checksum), "")
+		}
+	}
+
+	switch longhorn.BackingImageDataSourceType(backingImage.Spec.SourceType) {
+	case longhorn.BackingImageDataSourceTypeDownload:
+		if backingImage.Spec.SourceParameters[longhorn.DataSourceTypeDownloadParameterURL] == "" {
+			return werror.NewInvalidError(fmt.Sprintf("invalid parameter %+v for source type %v", backingImage.Spec.SourceParameters, backingImage.Spec.SourceType), "")
+		}
+	case longhorn.BackingImageDataSourceTypeUpload:
+	case longhorn.BackingImageDataSourceTypeExportFromVolume:
+		volumeName := backingImage.Spec.SourceParameters[controller.DataSourceTypeExportFromVolumeParameterVolumeName]
+		if volumeName == "" {
+			return werror.NewInvalidError(fmt.Sprintf("invalid parameter %+v for source type %v", backingImage.Spec.SourceParameters, backingImage.Spec.SourceType), "")
+		}
+		v, err := b.ds.GetVolume(volumeName)
+		if err != nil {
+			return werror.NewInvalidError(fmt.Sprintf("failed to get volume %v before exporting backing image", volumeName), "")
+		}
+		if v.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+			return werror.NewInvalidError(fmt.Sprintf("cannot export a backing image from faulted volume %v", volumeName), "")
+		}
+		eiName := types.GetEngineImageChecksumName(v.Status.CurrentImage)
+		ei, err := b.ds.GetEngineImage(eiName)
+		if err != nil {
+			return werror.NewInvalidError(fmt.Sprintf("failed to get then check engine image %v for volume %v before exporting backing image", eiName, volumeName), "")
+		}
+		if ei.Status.CLIAPIVersion < engineapi.CLIVersionFive {
+			return werror.NewInvalidError(fmt.Sprintf("engine image %v CLI version %v doesn't support this feature, please upgrade engine for volume %v before exporting backing image from the volume", eiName, ei.Status.CLIAPIVersion, volumeName), "")
+		}
+
+		if backingImage.Spec.SourceParameters[manager.DataSourceTypeExportFromVolumeParameterExportType] != manager.DataSourceTypeExportFromVolumeParameterExportTypeRAW &&
+			backingImage.Spec.SourceParameters[manager.DataSourceTypeExportFromVolumeParameterExportType] != manager.DataSourceTypeExportFromVolumeParameterExportTypeQCOW2 {
+			return werror.NewInvalidError(fmt.Sprintf("unsupported export type %v", backingImage.Spec.SourceParameters[manager.DataSourceTypeExportFromVolumeParameterExportType]), "")
+		}
+	default:
+		return werror.NewInvalidError(fmt.Sprintf("unknown backing image source type %v", backingImage.Spec.SourceType), "")
+	}
+
+	if _, err := b.ds.GetBackingImage(backingImage.Name); err == nil {
+		return werror.NewInvalidError("backing image already exists", "")
+	} else if !apierrors.IsNotFound(err) {
+		err = errors.Wrapf(err, "failed to check backing image existence before creation")
+		return werror.NewInvalidError(err.Error(), "")
+	}
```

Comment:
Is this redundant?

Reply:
Hmm?

Reply:
If the backing image CR already exists before the creation, the creation function will directly return an error, right?

Reply:
Good catch. It's redundant, because the creation function won't be executed if the backing image CR already exists.